[9.3.0-rc-2] NPE org.gradle.api.internal.artifacts.ivyservice...

### Current Behavior

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':firebase_remote_config:compileDebugKotlin'.
> Error while evaluating property 'friendPathsSet$kotlin_gradle_plugin_common' of task ':firebase_remote_config:compileDebugKotlin'.
   > Could not resolve all dependencies for configuration ':firebase_remote_config:debugCompileClasspath'.
      > Cannot invoke "org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder.SelectorState.getTargetModule()" because "this.sele
ctor" is null
```

### Expected Behavior

works well with rc-1 and all older versions

### Context (optional)

_No response_

### Self-contained Reproducer Project

flutter run
(happens on 2 different flutter projects)

### Gradle version

9.3.0-rc2

### Gradle version that used to work

9.3.0-rc1

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/EdgeState.java
+++ b/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/EdgeState.java
@@ -522,7 +522,7 @@ class EdgeState implements DependencyGraphEdge {
         this.unattached = true;
     }
 
-    public void markAttached() {
+    public void markNotUnattached() {
         this.unattached = false;
     }
 

--- a/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/ModuleResolveState.java
+++ b/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/ModuleResolveState.java
@@ -295,7 +295,7 @@ public class ModuleResolveState implements CandidateModule {
 
     public void removeUnattachedEdge(EdgeState edge) {
         if (unattachedEdges.remove(edge)) {
-            edge.markAttached();
+            edge.markNotUnattached();
         }
     }
 
@@ -402,8 +402,9 @@ public class ModuleResolveState implements CandidateModule {
     }
 
     void disconnectIncomingEdge(NodeState removalSource, EdgeState incomingEdge) {
-        incomingEdge.clearSelector();
+        // Remove the unattached edge first, as clearing the selector may trigger re-selection and mutate the unattached edge
         removeUnattachedEdge(incomingEdge);
+        incomingEdge.clearSelector();
         if (!incomingEdge.isConstraint()) {
             pendingDependencies.decreaseHardEdgeCount();
             if (pendingDependencies.isPending()) {

